webapp/crashstats/tests/test_libbugzilla.py

```diff
@@ -0,0 +1,677 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+from textwrap import dedent
+
+import pytest
+
+from crashstats.libbugzilla import (
+    crash_report_to_description,
+    minidump_thread_to_frames,
+    truncate,
+)
+
+
+@pytest.mark.parametrize(
+    "text, length, expected",
+    [
+        pytest.param("", 80, "", id="empty-string"),
+        pytest.param("abc123", 80, "abc123", id="basic-string"),
+        pytest.param("a" * 15, 10, ("a" * 7) + "...", id="truncated-string"),
+    ],
+)
+def test_truncate(text, length, expected):
+    assert truncate(text, length) == expected
+
+
+@pytest.mark.parametrize(
+    "thread, expected",
+    [
+        pytest.param(
```

Comment:
One problem with this test code is we don't know if it's valid data and we won't know if things change in the future such that it becomes invalid data at some point.

[Bug 1795699](https://bugzilla.mozilla.org/show_bug.cgi?id=1795699) covers implementing some bits to make it possible to validate test data as valid processed crash data. We can then use that in the tests to make sure that any test data we have doesn't become invalid in the future due to changes in the schema. Some day....

Reply:
I recall you and Sven talking about this a while back in another context, but another option that's more generalizable is to use static typing (I'm coming from TypeScript which is what FxA uses).

I was looking for the TypeScript analog, and Python does have some libraries and packages where you can define types, declare variables as those types and set up static type checking.

I know that'd be a lot more work to set up, but a test scenario like this seems like a good, low-risk way of introducing something like that to the code base.

Reply:
Ah okay. Yeah what I was thinking would replace however it's being represented now. It would mean we wouldn't have to write/maintain any custom validation code for example for this or any other schema.

Relatedly, I recall [using validation libraries](https://github.com/mozilla/fxa/blob/3354f25b44b06326bfa368950d8909fdfd4262d4/packages/fxa-auth-server/lib/routes/validators.js#L431) like [`joi`](https://joi.dev/api/?v=17.12.3) for complex/composable schemas in Node in FxA. Not sure if there are similar packages available in Python. Sounds like [`schematic`](https://schematics.readthedocs.io/en/latest/) [might be similar/well-liked](https://www.reddit.com/r/Python/comments/4xv0dk/python_validation_framework/).

Anyway, I'm not proposing any dramatic changes here; I'm just thinking out loud about different ways of accomplishing this kind of task.

Reply:
When I switched the processor to a schema-based system, I went with JSON schema defined schemas because that's what Data Org had standardized on and crash ingestion was part of Data Org and there was an impetus to unify how we were doing things with how they did things. After settling on a JSON schema-based system, I hunted for libraries and didn't find any Python libraries that did what I needed.

This is partially figured out in this bug:

https://bugzilla.mozilla.org/show_bug.cgi?id=1626698

After building out the schemas, I built some reducer and validation code that allows us to do permission-based reduction and some other things.

https://github.com/mozilla-services/socorro/tree/main/socorro/schemas

https://github.com/mozilla-services/socorro/blob/main/socorro/lib/libsocorrodataschema.py

Some of the processor and large portions of the webapp are built on top of that schema foundation, so switching at this point would be a big project.

I wrote a retrospective here:

https://bluesock.org/~willkg/blog/mozilla/socorro_schema_based_overhaul.html